### PR TITLE
Adds AT90USB162 support

### PR DESCRIPTION
**Pull request by daskygit**
## Description

Adds support for the AT90USB162.

After my previous PR failure I hope this one is smoother. I've ran `make all:default` and there are failures with some boards but they appear to happen on develop as well. I've compiled and flashed AT90USB162, ATmega32u4 and STM32F401 boards using this branch.

I've added preliminary support into spi_master, uart and serial but I've not tested these features but the registers, bits and pins line up with the datasheet. I've been able to test the changes to backlight_avr on a couple of accessible pins with hardware pwm. I'm a little unsure about the edit I've made to the max eeprom address in dynamic_keymap.c as it doesn't make an exception for the 16u2?

The MCU has no ADCs or support for I2C. I found a [datasheet](http://ww1.microchip.com/downloads/en/AppNotes/doc8224.pdf) while creating a the first pull request stating the ATmega16u2 is functional equivalent to the AT90USB162 which lines up with the edits I've made.

I'm expecting to have missed something but hopefully not too much. 

Edit:
For context this is to port existing hardware to QMK, I appreciate there's not likely going to be much call for this MCU. Any who thanks for your help and time.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by daskygit** [comment]
Also I forgot to ask, in common_features there's a comment saying ATmegaxxU2 doesn't have a hardware MUL instruction. 

When checking the avr instruction set they're listed as AVRe+ cores that should have the instruction with the only missing instructions being "ELPM, EIJMP, EICALL". It wouldn't be the first time the datasheets were wrong about something. 

I've added the AT90USB162 as a precaution considering the ATmega16u2 should be a replacement and I guess the only downside being performance.

File: quantum/backlight/backlight_avr.c [hunk 1]
```diff
@@ -68,7 +68,7 @@
 #        define COMxx1 COM3A1
 #        define OCRxx OCR3A
 #    endif
-#elif (defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega32U2__)) && (BACKLIGHT_PIN == B7 || BACKLIGHT_PIN == C5 || BACKLIGHT_PIN == C6)
+#elif (defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega32U2__) || defined(__AVR_AT90USB162__)) && (BACKLIGHT_PIN == B7 || BACKLIGHT_PIN == C5 || BACKLIGHT_PIN == C6)
```
Comment (fauxpark) [inline quantum/backlight/backlight_avr.c:71]:
```suggestion
#elif (defined(__AVR_AT90USB162__) || defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega32U2__)) && (BACKLIGHT_PIN == B7 || BACKLIGHT_PIN == C5 || BACKLIGHT_PIN == C6)
```

File: quantum/config_common.h [hunk 2]
```diff
@@ -39,7 +39,7 @@
 #        define PIND_ADDRESS 0x9
 #        define PINE_ADDRESS 0xC
 #        define PINF_ADDRESS 0xF
-#    elif defined(__AVR_ATmega32U2__) || defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega328P__) || defined(__AVR_ATmega328__)
+#    elif defined(__AVR_ATmega32U2__) || defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega328P__) || defined(__AVR_ATmega328__) || defined(__AVR_AT90USB162__)
```
Comment (fauxpark) [inline quantum/config_common.h:42]:
```suggestion
#    elif defined(__AVR_AT90USB162__) || defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega32U2__) || defined(__AVR_ATmega328P__) || defined(__AVR_ATmega328__)
```

File: quantum/mcu_selection.mk [hunk 3]
```diff
@@ -279,7 +279,7 @@ ifneq ($(findstring STM32F411, $(MCU)),)
   DFU_SUFFIX_ARGS ?= -v 0483 -p DF11
 endif
 
-ifneq (,$(filter $(MCU),atmega16u2 atmega32u2 atmega16u4 atmega32u4 at90usb646 at90usb647 at90usb1286 at90usb1287))
+ifneq (,$(filter $(MCU),atmega16u2 atmega32u2 atmega16u4 atmega32u4 at90usb646 at90usb647 at90usb1286 at90usb1287 at90usb162))
```
Comment (fauxpark) [inline quantum/mcu_selection.mk:282]:
```suggestion
ifneq (,$(filter $(MCU),at90usb162 atmega16u2 atmega32u2 atmega16u4 atmega32u4 at90usb646 at90usb647 at90usb1286 at90usb1287))
```

File: quantum/mcu_selection.mk [hunk 4]
```diff
@@ -317,7 +317,7 @@ ifneq (,$(filter $(MCU),atmega16u2 atmega32u2 atmega16u4 atmega32u4 at90usb646 a
   ifeq (,$(filter $(NO_INTERRUPT_CONTROL_ENDPOINT),yes))
     OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
   endif
-  ifneq (,$(filter $(MCU),atmega16u2 atmega32u2))
+  ifneq (,$(filter $(MCU),atmega16u2 atmega32u2 at90usb162))
```
Comment (fauxpark) [inline quantum/mcu_selection.mk:320]:
```suggestion
  ifneq (,$(filter $(MCU),at90usb162 atmega16u2 atmega32u2))
```

File: tmk_core/common/uart.c [hunk 5]
```diff
@@ -46,7 +46,7 @@
 #    define UDRIEn UDRIE0
 #    define USARTn_UDRE_vect USART_UDRE_vect
 #    define USARTn_RX_vect USART_RX_vect
-#elif defined(__AVR_ATmega32U4__) || defined(__AVR_ATmega32U2__) || defined(__AVR_AT90USB646__) || defined(__AVR_AT90USB647__) || defined(__AVR_AT90USB1286__) || defined(__AVR_AT90USB1287__)
+#elif defined(__AVR_ATmega32U4__) || defined(__AVR_ATmega32U2__) || defined(__AVR_AT90USB646__) || defined(__AVR_AT90USB647__) || defined(__AVR_AT90USB1286__) || defined(__AVR_AT90USB1287__) || defined(__AVR_AT90USB162__)
```
Comment (fauxpark) [inline tmk_core/common/uart.c:49]:
```suggestion
#elif defined(__AVR_AT90USB162__) || defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega32U2__) || defined(__AVR_ATmega16U4__) || defined(__AVR_ATmega32U4__) || defined(__AVR_AT90USB646__) || defined(__AVR_AT90USB647__) || defined(__AVR_AT90USB1286__) || defined(__AVR_AT90USB1287__)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: docs/feature_backlight.md [hunk 6]
```diff
@@ -93,18 +93,18 @@ BACKLIGHT_DRIVER = pwm
...
-|`B2`         |             |             |             |         |Timer 1    |
-|`B5`         |Timer 1      |Timer 1      |             |         |           |
-|`B6`         |Timer 1      |Timer 1      |             |         |           |
-|`B7`         |Timer 1      |Timer 1      |Timer 1      |         |           |
-|`C4`         |Timer 3      |             |             |         |           |
-|`C5`         |Timer 3      |             |Timer 1      |         |           |
-|`C6`         |Timer 3      |Timer 3      |Timer 1      |         |           |
-|`D4`         |             |             |             |Timer 1  |           |
-|`D5`         |             |             |             |Timer 1  |           |
+|Backlight Pin|AT90USB64/128|AT90USB162|ATmega16/32U4|ATmega16/32U2|ATmega32A|ATmega328/P|
```
[the remark this answers is not in the document]
Comment (daskygit) [reply docs/feature_backlight.md:96]: I couldn't decide on how to format it when adding to the existing column. Is there a preference for the field header?

So we have something like this below that looks good using `qmk docs` but a little harder to understand just as text.
```suggestion
|Backlight Pin|AT90USB64/128|ATmega16/32U4|AT90USB162<br>ATmega16/32U2|ATmega32A|ATmega328/P|
|-------------|-------------|-------------|-------------|---------|-----------|
|`B1`         |             |             |             |         |Timer 1    |
|`B2`         |             |             |             |         |Timer 1    |
|`B5`         |Timer 1      |Timer 1      |             |         |           |
|`B6`         |Timer 1      |Timer 1      |             |         |           |
|`B7`         |Timer 1      |Timer 1      |Timer 1      |         |           |
|`C4`         |Timer 3      |             |             |         |           |
|`C5`         |Timer 3      |             |Timer 1      |         |           |
|`C6`         |Timer 3      |Timer 3      |Timer 1      |         |           |
|`D4`         |             |             |             |Timer 1  |           |
|`D5`         |             |             |             |Timer 1  |           |
```


Or something like below but I think shortening it to 162 probably isn't clear enough and adding the full name makes the column extra wide.
```suggestion
|Backlight Pin|AT90USB64/128|ATmega16/32U4|ATmega16/32U2/162|ATmega32A|ATmega328/P|
|-------------|-------------|-------------|-----------------|---------|-----------|
|`B1`         |             |             |                 |         |Timer 1    |
|`B2`         |             |             |                 |         |Timer 1    |
|`B5`         |Timer 1      |Timer 1      |                 |         |           |
|`B6`         |Timer 1      |Timer 1      |                 |         |           |
|`B7`         |Timer 1      |Timer 1      |Timer 1          |         |           |
|`C4`         |Timer 3      |             |                 |         |           |
|`C5`         |Timer 3      |             |Timer 1          |         |           |
|`C6`         |Timer 3      |Timer 3      |Timer 1          |         |           |
|`D4`         |             |             |                 |Timer 1  |           |
|`D5`         |             |             |                 |Timer 1  |           |
```

There's probably another option but my markdown knowledge is non-existent.

File: docs/feature_backlight.md [hunk 7]
```diff
@@ -93,18 +93,18 @@ BACKLIGHT_DRIVER = pwm
...
+|`B1`         |             |          |             |             |         |Timer 1    |
+|`B2`         |             |          |             |             |         |Timer 1    |
+|`B5`         |Timer 1      |          |Timer 1      |             |         |           |
+|`B6`         |Timer 1      |          |Timer 1      |             |         |           |
+|`B7`         |Timer 1      |Timer 1   |Timer 1      |Timer 1      |         |           |
+|`C4`         |Timer 3      |          |             |             |         |           |
+|`C5`         |Timer 3      |Timer 1   |             |Timer 1      |         |           |
+|`C6`         |Timer 3      |Timer 1   |Timer 3      |Timer 1      |         |           |
+|`D4`         |             |          |             |             |Timer 1  |           |
+|`D5`         |             |          |             |             |Timer 1  |           |
```
Comment (daskygit) [inline docs/feature_backlight.md:107]: So we have something like this below that looks good using `qmk docs` but a little harder to understand just as text.
```suggestion
|Backlight Pin|AT90USB64/128|ATmega16/32U4|AT90USB162<br>ATmega16/32U2|ATmega32A|ATmega328/P|
|-------------|-------------|-------------|-------------|---------|-----------|
|`B1`         |             |             |             |         |Timer 1    |
|`B2`         |             |             |             |         |Timer 1    |
|`B5`         |Timer 1      |Timer 1      |             |         |           |
|`B6`         |Timer 1      |Timer 1      |             |         |           |
|`B7`         |Timer 1      |Timer 1      |Timer 1      |         |           |
|`C4`         |Timer 3      |             |             |         |           |
|`C5`         |Timer 3      |             |Timer 1      |         |           |
|`C6`         |Timer 3      |Timer 3      |Timer 1      |         |           |
|`D4`         |             |             |             |Timer 1  |           |
|`D5`         |             |             |             |Timer 1  |           |
```


Or something like below but I think shortening it to 162 probably isn't clear enough and adding the full name makes the column extra wide.
```suggestion
|Backlight Pin|AT90USB64/128|ATmega16/32U4|ATmega16/32U2/162|ATmega32A|ATmega328/P|
|-------------|-------------|-------------|-----------------|---------|-----------|
|`B1`         |             |             |                 |         |Timer 1    |
|`B2`         |             |             |                 |         |Timer 1    |
|`B5`         |Timer 1      |Timer 1      |                 |         |           |
|`B6`         |Timer 1      |Timer 1      |                 |         |           |
|`B7`         |Timer 1      |Timer 1      |Timer 1          |         |           |
|`C4`         |Timer 3      |             |                 |         |           |
|`C5`         |Timer 3      |             |Timer 1          |         |           |
|`C6`         |Timer 3      |Timer 3      |Timer 1          |         |           |
|`D4`         |             |             |                 |Timer 1  |           |
|`D5`         |             |             |                 |Timer 1  |           |
```

There's probably another option but my markdown knowledge is non-existent.

File: docs/feature_backlight.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (fauxpark) [reply docs/feature_backlight.md:96]: Yeah, that's fair. Might as well just leave it then :)

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by daskygit** [comment]
Resolved merge conflicts with uart.c and keyboard.jsonschema.

**Pull request merged**